File: rust/2021/src/bin/day_13.rs
use std::fs;
use std::collections::HashSet;

fn fold_grid(grid: &mut HashSet<(u32, u32)>, axis: char, val: u32) {
    for (x, y) in grid.clone() {
        if axis == 'x' && val < x {
            grid.remove(&(x, y));
            grid.insert((2 * val - x, y));
        } else if axis == 'y' && val < y {
            grid.remove(&(x, y));
            grid.insert((x, 2 * val - y));
        }
    }
}

fn main() {
    let input = fs::read_to_string("../input/2021/input_13.txt").unwrap();
    let (points, orders) = input.split_once("\n\n").unwrap();
    // Parse grid and instructions
    let mut grid: HashSet<(u32, u32)> = points.lines().map(|line| {
        let (x, y) = line.split_once(',').unwrap();
        (x.parse().unwrap(), y.parse().unwrap())
    }).collect();
    let folds: Vec<(char, u32)> = orders.lines().map(|line| {
        let (axis, val) = line.split_once('=').unwrap();
        (axis.chars().last().unwrap(), val.parse().unwrap())
    }).collect();
    // Perform first fold
    let mut fold_iter = folds.into_iter();
    let (axis, val) = fold_iter.next().unwrap();
    fold_grid(&mut grid, axis, val);
    println!("Part 1: {}", grid.len());
    // Perform the rest
    for (axis, val) in fold_iter {
        fold_grid(&mut grid, axis, val);
    }
    // Print resulting dots
    let x_max = *grid.iter().map(|(x, _)| x).max().unwrap();
    let y_max = *grid.iter().map(|(_, y)| y).max().unwrap();
    println!("Part 2:");
    for y in 0..=y_max {
        for x in 0..=x_max {
            print!("{}", if grid.contains(&(x, y)) { '#' } else { ' ' })
        }
        println!();
    }
}

File: haskell/2021/day_13.hs
import qualified Control.Monad as Monad
import qualified Data.Char as Char
import qualified Data.List.Split as Split
import qualified Data.Map as Map
import qualified Data.Set as Set

type Point = (Int, Int)

foldPoint :: (Char, Int) -> Point -> Point
foldPoint (axis, val) (x, y)
    | axis == 'x' && val < x = (2 * val - x, y)
    | axis == 'y' && val < y = (x, 2 * val - y)
    | otherwise = (x, y)

main :: IO ()
main = do
    text <- readFile "../../input/2021/input_13.txt"
    -- Parse grid and instructions
    let points:orders:_ = Split.splitOn "\n\n" text
    let grid = Set.fromList $ do
        line <- lines points
        let x:y:_ = read <$> Split.splitOn "," line :: [Int]
        return (x, y)
    let folds = do
        line <- lines orders
        let token = last $ words line
        let (axis:_):val:_ = Split.splitOn "=" token
        return (axis, read val :: Int)
    -- Do the first fold
    let singleFold = Set.map (foldPoint $ head folds) grid
    putStrLn $ "Part 1: " ++ show (Set.size singleFold)
    -- Do the rest
    let folded = foldl (\g f -> Set.map (foldPoint f) g) grid folds
    let (xMax, _) = Set.findMax folded
    let yMax = maximum $ snd <$> Set.elems folded
    putStrLn "Part 2:"
    Monad.forM_ [0..yMax] $ \y -> do
        Monad.forM_ [0..xMax] $ \x ->
            putChar $ if (x, y) `Set.member` folded then '#' else ' '
        putChar '\n'
